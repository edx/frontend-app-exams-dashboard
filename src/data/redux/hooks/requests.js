import { useSelector, useDispatch } from 'react-redux';

import * as redux from 'data/redux';
import * as module from './requests'; // eslint-disable-line import/no-self-import

const selectors = redux.selectors.requests;
const actions = redux.actions.requests;

export const statusSelector = selector => (requestName) => useSelector(selector(requestName));
export const useRequestIsPending = module.statusSelector(selectors.isPending);
export const useRequestIsFailed = module.statusSelector(selectors.isFailed);
export const useRequestIsCompleted = module.statusSelector(selectors.isCompleted);
export const useRequestIsInactive = module.statusSelector(selectors.isInactive);
export const useRequestError = module.statusSelector(selectors.error);
export const useRequestErrorCode = module.statusSelector(selectors.errorCode);
export const useRequestErrorStatus = module.statusSelector(selectors.errorStatus);
export const useRequestData = module.statusSelector(selectors.data);

export const useMakeNetworkRequest = () => {
  const dispatch = useDispatch();
  return ({
    requestKey,
    promise,
    onSuccess,
    onFailure,
  }) => {
    dispatch(actions.startRequest({ requestKey }));
    return promise.then((response) => {
      if (onSuccess) { onSuccess(response); }
      dispatch(actions.completeRequest({ requestKey, response }));
    }).catch((error) => {
      // Parse backend error from axios object
      let errorParsed = error;
      if (error?.customAttributes?.httpErrorResponseData) {
        errorParsed = JSON.parse(error.customAttributes.httpErrorResponseData);
      }
      if (onFailure) { onFailure(errorParsed); }
      dispatch(actions.failRequest({ requestKey, error: errorParsed }));
    });
  };
};

export const useClearRequest = (requestKey) => {
  const dispatch = useDispatch();
  return () => {
    dispatch(actions.clearRequest({ requestKey }));
  };
};
