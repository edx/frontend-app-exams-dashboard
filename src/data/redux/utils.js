import * as reduxHooks from 'data/redux/hooks';
import messages from 'pages/ExamsPage/messages';


const examAttemptStatusLabels = {
  created: messages.statusLabelCreated,
  download_software_clicked: messages.statusLabelDownloadSoftwareClicked,
  ready_to_start: messages.statusLabelReadyToStart,
  started: messages.statusLabelStarted,
  ready_to_submit: messages.statusLabelReadyToSubmit,
  submitted: messages.statusLabelSubmitted,
  verified: messages.statusLabelVerified,
  rejected: messages.statusLabelRejected,
  expired: messages.statusLabelExpired,
  second_review_required: messages.statusLabelSecondReviewRequired,
  error: messages.statusLabelError,
};

export const getMessageLabelForStatus = (status) => examAttemptStatusLabels[status];

export const getRequestStatusFromRedux = (requestKey) => {
  const isPending = reduxHooks.useRequestIsPending(requestKey);
  const isError = reduxHooks.useRequestError(requestKey);
  if (isPending) {
    return 'pending';
  } if (isError) {
    return 'error';
  }
  return '';
};