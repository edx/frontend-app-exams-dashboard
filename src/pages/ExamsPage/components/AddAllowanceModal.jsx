import React, { useState } from 'react';
import PropTypes from 'prop-types';
import {
  Alert,
  ModalDialog,
  Form,
  ActionRow,
  StatefulButton,
} from '@openedx/paragon';
import { Info } from '@openedx/paragon/icons';
import { useIntl } from '@edx/frontend-platform/i18n';

import * as constants from '../../../data/constants';
import { useClearRequest, useRequestError } from '../../../data/redux/hooks';
import { useFilteredExamsData, useCreateAllowance, useButtonStateFromRequestStatus } from '../hooks';
import messages from '../messages';

const AddAllowanceModal = ({ isOpen, close }) => {
  const { proctoredExams, timedExams } = useFilteredExamsData();

  const defaultExamType = proctoredExams.length > 0 ? 'proctored' : 'timed';
  const defaultExamsList = defaultExamType === 'proctored' ? proctoredExams : timedExams;

  const initialFormState = {
    'allowance-type': 'additional-minutes',
    'exam-type': defaultExamType,
    exams: {},
  };

  const [displayExams, setDisplayExams] = useState(defaultExamsList);
  const [form, setForm] = useState(initialFormState);
  const [learnerFieldError, setLearnerFieldError] = useState(false);
  const [examFieldError, setExamFieldError] = useState(false);
  const [additionalTimeError, setAdditionalTimeError] = useState(false);
  const createAllowanceRequestStatus = useButtonStateFromRequestStatus(constants.RequestKeys.createAllowance);
  const createAllowance = useCreateAllowance();
  const requestError = useRequestError(constants.RequestKeys.createAllowance);
  const resetRequestError = useClearRequest(constants.RequestKeys.createAllowance);
  const { formatMessage } = useIntl();

  const handleExamTypeChange = (examType) => {
    if (examType === 'proctored') {
      setDisplayExams(proctoredExams);
    } else {
      setDisplayExams(timedExams);
    }
    // reset exam selection state
    setForm(prev => ({ ...prev, exams: {} }));
  };

  const handleChange = (event) => {
    const { name, value } = event.target;
    resetRequestError();
    setForm(prev => ({ ...prev, [name]: value }));
    if (name === 'exam-type') {
      handleExamTypeChange(value);
    }
  };

  const handleCheckboxChange = (event) => {
    const el = event.target;
    const id = el.getAttribute('data-key');
    const timeLimitMins = el.value;
    const selectedExams = { ...form.exams };

    if (el.checked) {
      selectedExams[id] = timeLimitMins;
    } else {
      delete selectedExams[id];
    }
    setForm(prev => ({ ...prev, exams: selectedExams }));
  };

  const resetForm = () => {
    resetRequestError();
    setForm(initialFormState);
    setDisplayExams(defaultExamsList);
    setLearnerFieldError(false);
    setExamFieldError(false);
    setAdditionalTimeError(false);
  };

  const onClose = () => {
    resetForm();
    close();
  };

  const onSubmit = (e) => {
    e.preventDefault();
    setLearnerFieldError(!form.users);
    setExamFieldError(Object.keys(form.exams).length === 0);
    setAdditionalTimeError(!form['additional-time-minutes'] && !form['additional-time-multiplier']);
    const valid = (
      form.users
        && Object.keys(form.exams).length > 0
        && (form['additional-time-minutes'] || form['additional-time-multiplier'])
    );
    if (valid) {
      createAllowance(form, () => {
        setForm(initialFormState);
        onClose();
      });
    }
  };

  return (
    <ModalDialog
      title={formatMessage(messages.addAllowanceModalTitle)}
      isOpen={isOpen}
      onClose={onClose}
      size="md"
      hasCloseButton
      isFullscreenOnMobile
    >
      <ModalDialog.Header>
        <ModalDialog.Title>
          { formatMessage(messages.addAllowanceModalTitle) }
        </ModalDialog.Title>
      </ModalDialog.Header>
      <ModalDialog.Body>
        { requestError
          && (
          <Alert
            variant="danger"
            icon={Info}
          >
            <Alert.Heading>{ formatMessage(messages.addAllowanceFailedAlertHeader) }</Alert.Heading>
            <p>
              { requestError.detail }
            </p>
          </Alert>
          )}
        <Form id="add-allowance-form" onSubmit={onSubmit}>
          <Form.Group isInvalid={learnerFieldError}>
            <Form.Label>{ formatMessage(messages.addAllowanceLearnerField) }</Form.Label>
            <Form.Control name="users" value={form.users || ''} onChange={handleChange} data-testid="users" />
            { !learnerFieldError
              ? (
                <Form.Control.Feedback>
                  { formatMessage(messages.addAllowanceLearnerFieldFeedback) }
                </Form.Control.Feedback>
              )
              : (
                <Form.Control.Feedback type="invalid">
                  { formatMessage(messages.addAllowanceLearnerFieldErrorFeedback) }
                </Form.Control.Feedback>
              )}
          </Form.Group>
          <Form.Group controlId="form-exam-type">
            <Form.Label>{ formatMessage(messages.addAllowanceExamTypeField) }</Form.Label>
            <Form.Control
              as="select"
              onChange={handleChange}
              name="exam-type"
              value={form['exam-type']}
              data-testid="exam-type"
            >
              {
                proctoredExams.length > 0
                && <option value="proctored">{ formatMessage(messages.addAllowanceProctoredExamOption) }</option>
              }
              {
                timedExams.length > 0
                && <option value="timed">{ formatMessage(messages.addAllowanceTimedExamOption) }</option>
              }
            </Form.Control>
          </Form.Group>
          { displayExams.length > 0 ? (
            <Form.Group isInvalid={examFieldError}>
              <Form.Label>{ formatMessage(messages.addAllowanceExamField) }</Form.Label>
              <Form.CheckboxSet
                name="exams"
                onChange={handleCheckboxChange}
              >
                {
                  displayExams.map((exam) => (
                    <Form.Checkbox
                      key={exam.id}
                      data-key={exam.id}
                      value={exam.timeLimitMins}
                    >
                      {exam.name}
                    </Form.Checkbox>
                  ))
                }
              </Form.CheckboxSet>
              { examFieldError && <Form.Control.Feedback type="invalid">{ formatMessage(messages.addAllowanceExamErrorFeedback) }</Form.Control.Feedback>}
            </Form.Group>
          ) : null }
          <Form.Group controlId="form-allowance-type">
            <Form.Label>{ formatMessage(messages.addAllowanceAllowanceTypeField) }</Form.Label>
            <Form.Control
              as="select"
              name="allowance-type"
              onChange={handleChange}
              value={form['allowance-type']}
              data-testid="allowance-type"
            >
              <option value="additional-minutes">{ formatMessage(messages.allowanceAdditionalMinutesOption) }</option>
              <option value="time-multiplier">{ formatMessage(messages.addAllowanceTimeMultiplierOption) }</option>
            </Form.Control>
          </Form.Group>
          { form['allowance-type'] === 'additional-minutes'
            ? (
              <Form.Group controlId="form-allowance-value-minutes" isInvalid={additionalTimeError}>
                <Form.Label>{ formatMessage(messages.addAllowanceMinutesField) }</Form.Label>
                <Form.Control
                  name="additional-time-minutes"
                  value={form['additional-time-minutes'] || ''}
                  onChange={handleChange}
                  data-testid="additional-time-minutes"
                />
                { additionalTimeError && <Form.Control.Feedback type="invalid">{ formatMessage(messages.allowanceMinutesErrorFeedback) }</Form.Control.Feedback> }
              </Form.Group>
            )
            : (
              <Form.Group controlId="form-allowance-value-multiplier" isInvalid={additionalTimeError}>
                <Form.Label>{ formatMessage(messages.addAllowanceMultiplierField) }</Form.Label>
                <Form.Control
                  name="additional-time-multiplier"
                  value={form['additional-time-multiplier'] || ''}
                  onChange={handleChange}
                />
                {
                  additionalTimeError
                    ? (
                      <Form.Control.Feedback type="invalid">
                        { formatMessage(messages.addAllowanceMultiplierFeedback) }
                      </Form.Control.Feedback>
                    )
                    : (
                      <Form.Control.Feedback>
                        { formatMessage(messages.addAllowanceMultiplierErrorFeedback) }
                      </Form.Control.Feedback>
                    )
                }
              </Form.Group>
            )}
        </Form>
      </ModalDialog.Body>

      <ModalDialog.Footer>
        <ActionRow>
          <ModalDialog.CloseButton variant="tertiary" data-testid="close-modal">
            { formatMessage(messages.addAllowanceCloseButton) }
          </ModalDialog.CloseButton>
          <StatefulButton
            data-testid="create-allowance-stateful-button"
            state={createAllowanceRequestStatus()}
            labels={{
              default: formatMessage(messages.addAllowanceButtonDefaultLabel),
              pending: formatMessage(messages.addAllowanceButtonPendingLabel),
              complete: formatMessage(messages.addAllowanceButtonCompleteLabel),
              error: formatMessage(messages.addAllowanceButtonErrorLabel),
            }}
            type="submit"
            form="add-allowance-form"
          />
        </ActionRow>
      </ModalDialog.Footer>
    </ModalDialog>
  );
};

AddAllowanceModal.propTypes = {
  isOpen: PropTypes.bool.isRequired,
  close: PropTypes.func.isRequired,
};

export default AddAllowanceModal;
