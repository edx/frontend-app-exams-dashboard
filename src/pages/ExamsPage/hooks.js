import React from 'react';
import { useSelector, useDispatch } from 'react-redux';

import { RequestKeys } from 'data/constants';

import * as reduxHooks from 'data/redux/hooks';

import * as api from './data/api';
import * as selectors from './data/selectors';
import * as reducer from './data/reducer';

import * as module from './hooks'; // eslint-disable-line import/no-self-import

export const state = {
  // example of component state that does not need to be in redux
  exampleValue: (val) => React.useState(val), // eslint-disable-line
};

export const useFetchCourseExams = () => {
  const makeNetworkRequest = reduxHooks.useMakeNetworkRequest();
  const dispatch = useDispatch();
  return (courseId) => (
    makeNetworkRequest({
      requestKey: RequestKeys.fetchCourseExams,
      promise: api.getCourseExams(courseId),
      onSuccess: (response) => dispatch(reducer.loadExams(response)),
    })
  );
};

export const useFetchExamAttempts = () => {
  const makeNetworkRequest = reduxHooks.useMakeNetworkRequest();
  const dispatch = useDispatch();
  return (examId) => (
    makeNetworkRequest({
      requestKey: RequestKeys.fetchExamAttempts,
      promise: api.getExamAttempts(examId),
      onSuccess: (response) => dispatch(reducer.loadExamAttempts(response)),
    })
  );
};

export const useInitializeExamsPage = (courseId) => {
  const fetchCourseExams = module.useFetchCourseExams();
  React.useEffect(() => { fetchCourseExams(courseId); }, []); // eslint-disable-line react-hooks/exhaustive-deps
};

export const useExamsData = () => {
  const [exampleValue, setExampleValue] = state.exampleValue(0);
  const examsList = useSelector(selectors.courseExamsList);
  const currentExam = useSelector(selectors.currentExam);
  const isLoading = reduxHooks.useRequestIsPending(RequestKeys.fetchCourseExams);

  return {
    currentExam,
    examsList,
    isLoading,
    exampleValue,
    setExampleValue,
  };
};

export const useExamAttemptsData = () => {
  const attemptsList = useSelector(selectors.courseExamAttemptsList);
  return { attemptsList };
};
