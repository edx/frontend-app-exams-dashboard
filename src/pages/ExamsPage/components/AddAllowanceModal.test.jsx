import { render, screen, fireEvent } from '@testing-library/react';

import * as hooks from '../hooks';
import * as reduxHooks from '../../../data/redux/hooks';
import AddAllowanceModal from './AddAllowanceModal';

const mockMakeNetworkRequest = jest.fn();
const mockCreateAllowance = jest.fn();
const mockClearRequest = jest.fn();
const mockRequestError = jest.fn();

const proctoredExams = [
  { id: 1, name: 'exam1', examType: 'proctored', timeLimitMins: 60 }, // eslint-disable-line object-curly-newline
  { id: 3, name: 'exam3', examType: 'proctored', timeLimitMins: 45 }, // eslint-disable-line object-curly-newline
];

const timedExams = [{ id: 2, name: 'exam2', examType: 'timed', timeLimitMins: 30 }]; // eslint-disable-line object-curly-newline

// normally mocked for unit tests but required for rendering/snapshots
jest.unmock('react');

jest.mock('../hooks', () => ({
  useExamsData: jest.fn(),
  useFilteredExamsData: jest.fn(),
  useButtonStateFromRequestStatus: jest.fn(),
  useCreateAllowance: jest.fn(),
}));

jest.mock('../../../data/redux/hooks', () => ({
  useRequestError: jest.fn(),
  useClearRequest: jest.fn(),
}));

describe('AddAllowanceModal', () => {
  beforeEach(() => {
    jest.resetAllMocks();
    hooks.useFilteredExamsData.mockReturnValue({ proctoredExams, timedExams });
    hooks.useButtonStateFromRequestStatus.mockReturnValue(mockMakeNetworkRequest);
    hooks.useCreateAllowance.mockReturnValue(mockCreateAllowance);
    reduxHooks.useRequestError.mockReturnValue(mockRequestError);
    reduxHooks.useClearRequest.mockReturnValue(mockClearRequest);
  });

  it('should match snapshot', () => {
    expect(render(<AddAllowanceModal isOpen close={jest.fn()} />)).toMatchSnapshot();
  });

  it('should display timed exam choices', () => {
    render(<AddAllowanceModal isOpen close={jest.fn()} />);
    fireEvent.change(screen.getByTestId('exam-type'), { target: { value: 'timed' } });
    expect(screen.getByText('exam2')).toBeInTheDocument();
  });

  it('should update allowance input', () => {
    render(<AddAllowanceModal isOpen close={jest.fn()} />);
    fireEvent.change(screen.getByTestId('allowance-type'), { target: { value: 'time-multiplier' } });
    expect(screen.getByText('Multiplier')).toBeInTheDocument();
  });

  it('should submit with default values', () => {
    render(<AddAllowanceModal isOpen close={jest.fn()} />);
    fireEvent.change(screen.getByTestId('users'), { target: { value: 'edx, edx@example.com' } });
    fireEvent.click(screen.getByText('exam1'));
    fireEvent.click(screen.getByText('exam3'));
    fireEvent.change(screen.getByTestId('additional-time-minutes'), { target: { value: '60' } });
    fireEvent.click(screen.getByTestId('create-allowance-stateful-button'));
    const expectedData = {
      users: 'edx, edx@example.com',
      'exam-type': 'proctored',
      exams: { 1: '60', 3: '45' },
      'additional-time-minutes': '60',
      'allowance-type': 'additional-minutes',
    };
    expect(mockCreateAllowance).toHaveBeenCalledWith(expectedData, expect.any(Function));
  });

  it('should display field errors', () => {
    render(<AddAllowanceModal isOpen close={jest.fn()} />);
    fireEvent.click(screen.getByTestId('create-allowance-stateful-button'));
    expect(screen.getByText('Enter learners')).toBeInTheDocument();
    expect(screen.getByText('Select exams')).toBeInTheDocument();
    expect(screen.getByText('Enter minutes greater than 0')).toBeInTheDocument();
  });

  it('should show an alert if the request fails', () => {
    reduxHooks.useRequestError.mockReturnValue({ detail: 'some test error' });
    render(<AddAllowanceModal isOpen close={jest.fn()} />);
    expect(screen.getByText('some test error')).toBeInTheDocument();
  });

  it('should clear request errors if form is altered', () => {
    render(<AddAllowanceModal isOpen close={jest.fn()} />);
    fireEvent.change(screen.getByTestId('users'), { target: { value: 'edx, edx@example.com' } });
    expect(reduxHooks.useClearRequest).toHaveBeenCalled();
  });

  it('should reset form state when closed', () => {
    render(<AddAllowanceModal isOpen close={jest.fn()} />);
    fireEvent.change(screen.getByTestId('exam-type'), { target: { value: 'timed' } });
    fireEvent.click(screen.getByTestId('create-allowance-stateful-button'));

    expect(screen.getByText('Enter minutes greater than 0')).toBeInTheDocument();
    expect(screen.getByText('exam2')).toBeInTheDocument();

    fireEvent.click(screen.getByTestId('close-modal'));
    // any errors and timed exam selections should be reset
    expect(screen.queryByText('Enter minutes greater than 0')).not.toBeInTheDocument();
    expect(screen.queryByText('exam2')).not.toBeInTheDocument();
    // expect the default selection to revert back to proctored exams
    expect(screen.queryByText('exam1')).toBeInTheDocument();
  });
});
