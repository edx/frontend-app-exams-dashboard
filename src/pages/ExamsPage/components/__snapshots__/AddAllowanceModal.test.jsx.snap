// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddAllowanceModal should match snapshot 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body
    class="pgn__hidden-scroll-padding-right"
    data-scroll-locked="1"
  >
    <div
      aria-hidden="true"
      data-focus-on-hidden="true"
    />
    <div
      id="paragon-portal-root"
    >
      <div
        aria-hidden="true"
        data-focus-guard="true"
        data-focus-on-hidden="true"
        style="width: 1px; height: 0px; padding: 0px; overflow: hidden; position: fixed; top: 1px; left: 1px;"
        tabindex="0"
      />
      <div
        class="pgn__modal-layer"
        data-focus-lock-disabled="false"
      >
        <div
          class="pgn__modal-content-container"
        >
          <div
            class="pgn__modal-backdrop"
            data-testid="modal-backdrop"
          />
          <div
            aria-label="Add a new allowance"
            class="pgn__modal pgn__modal-md pgn__modal-default pgn__modal-visible-overflow"
            role="dialog"
          >
            <div
              class="pgn__modal-close-container"
            >
              <button
                aria-label="Close"
                class="btn-icon btn-icon-primary btn-icon-md pgn__modal-close-button"
                type="button"
              >
                <span
                  class="btn-icon__icon-container"
                >
                  <span
                    class="pgn__icon btn-icon__icon"
                  >
                    <svg
                      aria-hidden="true"
                      fill="none"
                      focusable="false"
                      height="24"
                      role="img"
                      viewBox="0 0 24 24"
                      width="24"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <path
                        d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12 19 6.41Z"
                        fill="currentColor"
                      />
                    </svg>
                  </span>
                </span>
              </button>
            </div>
            <div
              class="pgn__modal-header"
            >
              <h2
                class="pgn__modal-title"
              >
                Add a new allowance
              </h2>
            </div>
            <div
              class="pgn__modal-body pgn__modal-body-scroll-top pgn__modal-body-scroll-bottom"
            >
              <div />
              <div
                class="pgn__modal-body-content"
              >
                <div
                  class="fade alert-content alert alert-danger show"
                  role="alert"
                >
                  <span
                    class="pgn__icon alert-icon"
                  >
                    <svg
                      aria-hidden="true"
                      fill="none"
                      focusable="false"
                      height="24"
                      role="img"
                      viewBox="0 0 24 24"
                      width="24"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <path
                        d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2Zm1 15h-2v-6h2v6Zm0-8h-2V7h2v2Z"
                        fill="currentColor"
                      />
                    </svg>
                  </span>
                  <div
                    class="pgn__alert-message-wrapper"
                  >
                    <div
                      class="alert-message-content"
                    >
                      <div
                        class="alert-heading h4"
                      >
                        We couldn't apply your changes
                      </div>
                      <p />
                    </div>
                  </div>
                </div>
                <form
                  class=""
                  id="add-allowance-form"
                >
                  <div
                    class="pgn__form-group"
                  >
                    <label
                      class="pgn__form-label"
                      for="form-field1"
                    >
                      Learners
                    </label>
                    <div
                      class="pgn__form-control-decorator-group"
                    >
                      <input
                        aria-describedby="form-field1-6"
                        class="form-control"
                        data-testid="users"
                        id="form-field1"
                        name="users"
                        value=""
                      />
                    </div>
                    <div
                      class="pgn__form-control-description pgn__form-text pgn__form-text-default"
                      id="form-field1-6"
                    >
                      <div>
                        Enter usernames or emails separated by commas
                      </div>
                    </div>
                  </div>
                  <div
                    class="pgn__form-group"
                  >
                    <label
                      class="pgn__form-label"
                      for="form-exam-type"
                    >
                      Exam type
                    </label>
                    <div
                      class="pgn__form-control-decorator-group"
                    >
                      <select
                        class="has-value form-control"
                        data-testid="exam-type"
                        id="form-exam-type"
                        name="exam-type"
                      >
                        <option
                          value="proctored"
                        >
                          Proctored exam
                        </option>
                        <option
                          value="timed"
                        >
                          Timed exam
                        </option>
                      </select>
                    </div>
                  </div>
                  <div
                    class="pgn__form-group"
                  >
                    <p
                      class="pgn__form-label"
                      id="form-field2-8"
                    >
                      Exams
                    </p>
                    <div
                      aria-labelledby="form-field2-8"
                      class="pgn__form-control-set"
                      id="form-field2"
                      role="group"
                    >
                      <div
                        class="pgn__form-checkbox"
                      >
                        <input
                          class="pgn__form-checkbox-input"
                          data-key="1"
                          id="form-field3"
                          name="exams"
                          type="checkbox"
                          value="60"
                        />
                        <div>
                          <label
                            class="pgn__form-label"
                            for="form-field3"
                          >
                            exam1
                          </label>
                        </div>
                      </div>
                      <div
                        class="pgn__form-checkbox"
                      >
                        <input
                          class="pgn__form-checkbox-input"
                          data-key="3"
                          id="form-field4"
                          name="exams"
                          type="checkbox"
                          value="45"
                        />
                        <div>
                          <label
                            class="pgn__form-label"
                            for="form-field4"
                          >
                            exam3
                          </label>
                        </div>
                      </div>
                    </div>
                  </div>
                  <div
                    class="pgn__form-group"
                  >
                    <label
                      class="pgn__form-label"
                      for="form-allowance-type"
                    >
                      Allowance type
                    </label>
                    <div
                      class="pgn__form-control-decorator-group"
                    >
                      <select
                        class="has-value form-control"
                        data-testid="allowance-type"
                        id="form-allowance-type"
                        name="allowance-type"
                      >
                        <option
                          value="additional-minutes"
                        >
                          Additional time (minutes)
                        </option>
                        <option
                          value="time-multiplier"
                        >
                          Time multiplier
                        </option>
                      </select>
                    </div>
                  </div>
                  <div
                    class="pgn__form-group"
                  >
                    <label
                      class="pgn__form-label"
                      for="form-allowance-value-minutes"
                    >
                      Minutes
                    </label>
                    <div
                      class="pgn__form-control-decorator-group"
                    >
                      <input
                        class="form-control"
                        data-testid="additional-time-minutes"
                        id="form-allowance-value-minutes"
                        name="additional-time-minutes"
                        value=""
                      />
                    </div>
                  </div>
                </form>
              </div>
              <div />
            </div>
            <div
              class="pgn__modal-footer"
            >
              <div
                class="pgn__action-row"
              >
                <button
                  class="pgn__modal-close-button btn btn-tertiary"
                  type="button"
                >
                  Cancel
                </button>
                <button
                  aria-disabled="false"
                  aria-live="assertive"
                  class="pgn__stateful-btn pgn__stateful-btn-state-default btn btn-primary"
                  data-testid="create-allowance-stateful-button"
                  form="add-allowance-form"
                  type="submit"
                >
                  <span
                    class="d-flex align-items-center justify-content-center"
                  >
                    <span>
                      Add allowance
                    </span>
                  </span>
                </button>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        aria-hidden="true"
        data-focus-guard="true"
        data-focus-on-hidden="true"
        style="width: 1px; height: 0px; padding: 0px; overflow: hidden; position: fixed; top: 1px; left: 1px;"
        tabindex="0"
      />
    </div>
  </body>,
  "container": <div
    aria-hidden="true"
    data-focus-on-hidden="true"
  />,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
